Elly Vendingmachine step1

dely2p opened the pull request.

dely2p committed:
create Beverage protocol #1

우유, 탄산음료, 커피의 공통된 필수 속성을 묶어둘 Beverage protocol을 만들었다.

dely2p committed:
create Milk, Soda, Coffee class #1

Beverage protocol을 사용하는 Milk, Soda, Coffee 클래스를 생성했다. 후에 요구사항에서 각각의 특징을 구분짓는 속성과 행동을 추가하려고 한다.

dely2p committed:
add menu property of enum type at each class #1

Milk, Soda, Coffee Class 각각에 Menu라는 enum type의 menu 속성을 추가했다. enum에는 요구사항에서 예시로 주어진 각 음료들의 이름을 넣었다. 또 생성자로 Menu type을 생성자로 받도록 수정했다.

dely2p committed:
add CustomStringConvertible at Milk, Soda, Coffee Class to print function #2

CustomStringConvertible protocol을 사용하여 음료 상품 클래스 출력 기능을 만들었다.

dely2p committed:
fix from Beverage protocol to Beverage Class #4

계층 구조인 자판기 특성으로protocol이 아닌 class 상속형태로 바꾸었다.

dely2p committed:
create a drink object instance by type in the main and output as a print() function #3

Date format을 yyyyMMdd형태로 정하여 음료수 객체를 만들고 Beverage의 CustomStringConvertible protocol의 description을 override하여 출력문을 만들었다.

dely2p committed:
fix from optional binding of beverage optional binding to nil coalescing optional #3

Date format 때문에 옵셔널 바인딩으로 만들었던 것을 nil coalescing optional을 사용하여 간략하게 표현하도록 개선했다.

godrm commented:
오 깔끔하게 잘 하셨네요. 상속을 활용한 객체 지향 프로그래밍 연습을 의도적으로 많이 해보기 바랍니다.
딱 하나만 더 도전해보죠! 
- 각 하위 클래스에 있는 `enum Menu`를 protocol 과 associatedtype (제네릭 타입)으로 추상화해보세요!

dely2p committed:
fix to abstract the enum menu in each subclass into protocol and associatedtype #5

MilkMenu, SodaMenu, CoffeeMenu 각각의 구조가 같으므로 associatedtype을 사용해서 제네릭 프로토콜로 만들어서 추상화했다.

godrm commented:
잘하셨습니다. 학습하면서 이해가 안되는 부분들을 꾸준히 정리해서 자기껄로 만들어가세요!
